bot/cogs/moderation/incidents.py

```diff
@@ -42,6 +42,19 @@ class Signal(Enum):
 ALL_SIGNALS: t.Set[str] = {signal.value for signal in Signal}
 
 
+async def download_file(attachment: discord.Attachment) -> t.Optional[discord.File]:
+    """
+    Download & return `attachment` file.
+
+    If the download fails, the reason is logged and None will be returned.
+    """
+    log.debug(f"Attempting to download attachment: {attachment.filename}")
+    try:
+        return await attachment.to_file()
+    except Exception:
+        log.exception("Failed to download attachment")
```

Comment:
`NotFound` shouldn't have a high level.

Reply:
Reduced to `debug` in 6f5fb20, though I still don't think I fully understand why a 404 is in this case different to any other error.

Here's the way I see it: when fetching messages, a 404 is expected in the case of a racing coroutine trying to get its hands on an already relayed message. It signals a certain situation that must be handled differently. An error notification isn't desired, because a) the situation is expected, and b) the behaviour for handling it is defined.

When downloading the attachment, however, I don't see a 404 as something that we should ignore. I cannot think of a way where this fails on a 404 and we don't want to know. I can trigger the 404 by putting a sleep before the download, and manually deleting the original message before `to_file` is called. But if this somehow gives a 404 in production, I feel like we should know?

Either way, it's not too important, just trying to explain my thought process so that you can correct me if I'm making an error somewhere in my thinking.

Reply:
If you knew there's a 404, then what exactly would you do with that information? You can't bring back a deleted file. We shouldn't have high logging levels for errors which aren't actionable. I assumed a 404 was expected if a message was manually deleted. Turns out it's a 403, but the same point still applies. You could argue that it'd be better to detect the message delete and abort before trying to get the file, but this may end up being too over-engineered compared to simply ignoring an exception. Even if it's a 403, a 404 is still worthwhile to ignore since, again, it's unactionable. It's just like all the 500s we ignore in Sentry. If a file is gone, then it's gone.

Reply:
> If you knew there's a 404, then what exactly would you do with that information? You can't bring back a deleted file.

Yes, but it just wasn't clear to me how a 404 is different from any other exception, because the original message gets deleted regardless. So if it's a 404 we lose the file but don't get an alert, if it's a 500 we lose the file but get an alert.

So do you think we need to be ready for the situation where someone actions an incident and deletes it at the same time? I guess I'll make it not log the exception if it's a 403 then too.

Reply:
Other exceptions may be the result of buggy code. But a 404 comes from a deleting a message, which is a normal user action.

Reply:
> So do you think we need to be ready for the situation where someone actions an incident and deletes it at the same time?

Not really. Things won't break catastrophically will they? In the case it does happen, I would just rather not be pestered by the 404/403 exception.

Reply:
> Things won't break catastrophically will they?

Depends, it's unfortunate. The thing is that the relay is pretty fast, unless it's downloading a massive file. You'd have to *try* to break it, it's extremely unlikely that it'd happen by accident. But if it does, then:

* Attachment fails to download, we fallback on `proxy_url`
* Let's delete, `confirmation_task` is scheduled
* The call to `message.delete` fails with an unhandled 404
* The `confirmation_task` is already running but isn't awaited in the try block, so the timeout is unhandled after 5 secs

So if someone deletes the message we'll get alerts anyway.

Reply:
I haven't though about it much, but I suppose the way to handle it would be to catch exceptions from the delete call and if we get something, then cancel `confirmation_task`? But it still feels fragile. I don't think we normally write code to be resilient to intentional malice. For what this cog does, it already feels over-engineered.

Reply:
Well, if you want to account for it, feel free. But I'm fine with revisiting this later if we actually do encounter a deleted message in production.

Reply:
I don't think we'll encounter it personally. I volunteer to handle that Sentry ticket if it ever comes through. :smile: 

403 logs at `debug` level too now: 476d407